src/middlewared/middlewared/plugins/pool_/dataset_encryption_info.py

```diff
@@ -308,6 +308,75 @@ def export_keys(self, job, id):
 
+    @accepts(Int('id'))
+    @returns()
+    @job(lock='replication_task_dataset_export_keys', pipes=['output'])
+    def export_keys_for_replication(self, job, task_id):
+        """
+        Export keys for `id` and its children which are stored in the system. The exported file is a JSON file
+        which has a dictionary containing dataset names as keys and their keys as the value.
+
+        Please refer to websocket documentation for downloading the file.
+        """
+        datasets = self.middleware.call_sync('replication.query_encrypted_roots_keys', task_id)
+        with BytesIO(json.dumps(datasets).encode()) as f:
+            shutil.copyfileobj(f, job.pipes.output.w)
+
+    @private
+    async def export_keys_for_replication_internal(self, replication_task_id):
+        task = await self.middleware.call('replication.get_instance', replication_task_id)
+        if task['direction'] != 'PUSH':
+            raise CallError('Only push replication tasks are supported.', errno.EINVAL)
+
+        await (await self.middleware.call(
+            'core.bulk', 'pool.dataset.sync_db_keys', [[source] for source in task['source_datasets']]
+        )).wait()
+
+        mapping = {}
+        for source_ds in task['source_datasets']:
+            if task['recursive']:
+                filters = ['OR', [['name', '=', source_ds], ['name', '^', f'{source_ds}/']]]
+            else:
+                filters = ['name', '=', source_ds]
+            mapping[source_ds] = await self.middleware.call('pool.dataset.query_encrypted_roots_keys', [filters])
+
+        # We have 3 cases to deal with
+        # 1. There are no encrypted datasets in source dataset, so let's just skip in that case
+        # 2. There is only 1 source dataset, in this case the destination dataset will be overwritten as is, so we
+        #    generate mapping accordingly. For example if source is `tank/enc` and destination is `dest/enc`, in
+        #    the destination system `dest/enc` will reflect `tank/enc` so we reflect that accordingly in the mapping
+        # 3. There are multiple source datasets, in this case they will become child of destination dataset
+
+        if not any(mapping.values()):
+            return {}
+
+        normalized_result = {}
+        include_encryption_root_children = not task['replicate'] and task['recursive']
+        target_ds = task['target_dataset']
+        if len(mapping) == 1:
+            source_ds = task['source_datasets'][0]
+            for ds_name, key in mapping[source_ds].items():
+                for dataset in await self.middleware.call(
+                    'pool.dataset.query', [['encryption_root', '=', ds_name]], {
+                        'extra': {'properties': ['encryptionroot']}
+                    }
+                ) if include_encryption_root_children else [{'id': ds_name}]:
+                    normalized_result[dataset['id'].replace(source_ds, target_ds, 1)] = key
```

Comment:
It is a little bit complicated than that
```
from zettarepl.replication.task.dataset import get_target_dataset

fake_replication_task = types.SimpleNamespace()
fake_replication_task.source_datasets = source_datasets
fake_replication_task.target_dataset = target_dataset

get_target_dataset(fake_replication_task, source_dataset)
```

(please isolate this in zettarepl.py so we don't import from zettarepl in other plugins)